### PR TITLE
aarch64: more MMU code improvements

### PR DESCRIPTION
**Pull request by npitre**
Various MMU changes that I'd like to upstream before the next batch.

- aarch64: mmu: dynamic mapping support
- aarch64: mmu: make page table pool global
- aarch64: mmu: move get_region_desc() closer to usage points
- aarch64: mmu: factor out the phys argument out of set_mapping()
- aarch64: mmu: invert the MT_OVERWRITE flag
- aarch64: mmu: use a range (start..end) for common MMU mapping
- aarch64: mmu: fix common MMU mapping
- aarch64: mmu: rationalize debugging output
- arm64: mmu: initialize early

### CODE REVIEW

**Comment by carlocaione** [comment]
Not something to fix but something to take into account: now that the MMU is initialized at really early boot (and correctly so) there is no way we can get any debug in output without resorting to semi-hosting. We should probably implement on top of this way of debugging something similar to what's done for X86 where an init routine is called to recursively dump all the pages.

**Review by carlocaione — APPROVED** [review]

**Comment by carlocaione** [comment]
@Abhishek-brcm @sandeepbrcm @ioannisg can you take a look at this? This is a blocking PR for the userspace work.

**Comment by ioannisg** [comment]
> @Abhishek-brcm @sandeepbrcm @ioannisg can you take a look at this? This is a blocking PR for the userspace work.

I will take a look as soon as possible

**Review by MrVan — APPROVED** [review]

File: arch/arm/core/aarch64/mmu/arm_mmu.c [hunk 1]
```diff
@@ -244,72 +228,206 @@ static int add_map(struct arm_mmu_ptables *ptables, const char *name,
...
+	uint32_t attrs;
+};
+
+static const struct arm_mmu_flat_range mmu_zephyr_ranges[] = {
 
 	/* Mark the zephyr execution regions (data, bss, noinit, etc.)
 	 * cacheable, read-write
 	 * Note: read-write region is marked execute-never internally
 	 */
-	MMU_REGION_FLAT_ENTRY("zephyr_data",
```
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/mmu/arm_mmu.c:278]: The definitions `MMU_REGION_FLAT_ENTRY  `can be removed from header then.

File: arch/arm/core/aarch64/mmu/arm_mmu.c [hunk 2]
```diff
@@ -19,89 +19,57 @@
...
 
-	if (ptables->next_table >= CONFIG_MAX_XLAT_TABLES) {
-		LOG_ERR("CONFIG_MAX_XLAT_TABLES, too small");
-		return NULL;
+	/* Look for a free table. */
+	for (i = 0; i < CONFIG_MAX_XLAT_TABLES; i++) {
+		if (xlat_use_count[i] == 0) {
+			xlat_use_count[i] = 1;
+			return &xlat_tables[i * Ln_XLAT_NUM_ENTRIES];
+		}
```
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/mmu/arm_mmu.c:36]: xlat_use_count[]  needs locking.

File: arch/arm/core/aarch64/mmu/arm_mmu.c [hunk 3]
```diff
@@ -498,3 +610,12 @@ void arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flags)
 		k_panic();
 	}
 }
+
+void arch_mem_unmap(void *addr, size_t size)
+{
+	int ret = remove_map(&kernel_ptables, "generic", (uintptr_t)addr, size);
+
```
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/mmu/arm_mmu.c:617]: needs `TLBI` . 
`arch_mem_map`  needs them too.
and the `enable_mmu_el1`  as well.

**Comment by npitre** [comment]
On Thu, 25 Feb 2021, Sandeep Tripathy wrote:

> >  /* zephyr execution regions with appropriate attributes */
> -static const struct arm_mmu_region mmu_zephyr_regions[] = {
> +
> +struct arm_mmu_flat_range {
> 
> static

Nope. That's a structure definition, not the actual declaration.

> The definitions `MMU_REGION_FLAT_ENTRY  `can be removed from header then.

Nope. It is still used in platform files where the start+size is 
appropriate, and used with the structure format that allows for a 
different physical address.

> > +	/* Look for a free table. */
> +	for (i = 0; i < CONFIG_MAX_XLAT_TABLES; i++) {
> +		if (xlat_use_count[i] == 0) {
> +			xlat_use_count[i] = 1;
> +			return &xlat_tables[i * Ln_XLAT_NUM_ENTRIES];
> +		}
> 
> xlat_use_count[]  needs locking.

Not at this point. Everything is either single-threaded, or serialized 
on z_mm_lock already.

More locking will come with a second batch of patches after this lot 
gets merged though.

> > @@ -498,3 +610,12 @@ void arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flags)
>  		k_panic();
>  	}
>  }
> +
> +void arch_mem_unmap(void *addr, size_t size)
> +{
> +	int ret = remove_map(&kernel_ptables, "generic", (uintptr_t)addr, size);
> +
> 
> needs `TLBI` . 
> `arch_mem_map`  needs them too.
> and the `enable_mmu_el1`  as well.

That, too, is waiting in another batch. Right now we set up the tables 
and never touch them again. Arguably in `enable_mmu_el1` this might be 
needed, but things will get much more complex later.

**Comment by sandeepbrcm** [comment]
> serialized on z_mm_lock already

I see

> That, too, is waiting in another batch

Sure thanks. I was just ensuring it is considered since this patch added dynamic map support (the APIs) regardless if someone is using it or not the TLB invalidations becomes an essential part of functional requirement. ie: the dynamic mapping functionality  is not supposed to work without TLBI in place. 

>  `enable_mmu_el1` this might be needed

Certainly.

**Review by sandeepbrcm — APPROVED** [review]
I was not able to review the impact of commit altering `__kernel_ram_xxx -> _image_ram_xxx`
(which I will understand later)
Rest of the patch looks good.

**Comment by npitre** [comment]
On Fri, 26 Feb 2021, Sandeep Tripathy wrote:

> I was not able to review the impact of commit altering `__kernel_ram_xxx -> _image_ram_xxx` 
> (which I will understand later)

This is relevant for memory domains introduced later, as well as dynamic 
virtual memory allocation whose CI test expects ummapped areas below 
`__kernel_ram_end`.

> Rest of the patch looks good. 

Great, thanks. More is to come.

**Pull request merged**